File: mobile/android/app/src/main/java/com/tigerwallet/app/services/mpc/MPCWalletService.java
package com.tigerwallet.app.services.mpc;

import android.content.Context;
import android.util.Base64;
import org.json.JSONArray;
import org.json.JSONObject;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.SecureRandom;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

/**
 * MPC Wallet Service for Android
 * Provides distributed key generation and transaction signing
 */
public class MPCWalletService {
    private static final String TAG = "MPCWalletService";
    private static final String BASE_URL = "https://api.tigerwallet.com/v1/mpc";
    
    private final Context context;
    private final ExecutorService executor;
    private String sessionKey;
    private Map<String, String> walletCache;
    
    public interface Callback<T> {
        void onSuccess(T result);
        void onError(Exception error);
    }
    
    public MPCWalletService(Context context) {
        this.context = context;
        this.executor = Executors.newFixedThreadPool(4);
        this.walletCache = new HashMap<>();
    }
    
    /**
     * Create a new MPC wallet
     */
    public void createWallet(String userId, Callback<MPCWallet> callback) {
        executor.execute(() -> {
            try {
                JSONObject body = new JSONObject();
                body.put("userId", userId);
                
                JSONObject response = makePostRequest("/wallet", body);
                
                MPCWallet wallet = new MPCWallet();
                wallet.address = response.getString("address");
                wallet.publicKey = response.getString("publicKey");
                wallet.createdAt = response.getLong("createdAt");
                
                walletCache.put(wallet.address, wallet.toJson());
                
                callback.onSuccess(wallet);
            } catch (Exception e) {
                callback.onError(e);
            }
        });
    }
    
    /**
     * Import existing wallet using mnemonic
     */
    public void importWallet(String mnemonic, String userId, Callback<MPCWallet> callback) {
        executor.execute(() -> {
            try {
                JSONObject body = new JSONObject();
                body.put("mnemonic", mnemonic);
                body.put("userId", userId);
                
                JSONObject response = makePostRequest("/wallet/import", body);
                
                MPCWallet wallet = new MPCWallet();
                wallet.address = response.getString("address");
                wallet.publicKey = response.getString("publicKey");
                wallet.createdAt = response.getLong("createdAt");
                
                walletCache.put(wallet.address, wallet.toJson());
                
                callback.onSuccess(wallet);
            } catch (Exception e) {
                callback.onError(e);
            }
        });
    }
    
    /**
     * Get public key for an address
     */
    public void getPublicKey(String address, Callback<String> callback) {
        executor.execute(() -> {
            try {
                JSONObject response = makeGetRequest("/publickey/" + address);
                callback.onSuccess(response.getString("publicKey"));
            } catch (Exception e) {
                callback.onError(e);
            }
        });
    }
    
    /**
     * Sign a transaction using MPC
     */
    public void signTransaction(String txData, String address, Callback<String> callback) {
        executor.execute(() -> {
            try {
                JSONObject body = new JSONObject();
                body.put("txData", txData);
                body.put("address", address);
                
                JSONObject response = makePostRequest("/sign", body);
                callback.onSuccess(response.getString("signature"));
            } catch (Exception e) {
                callback.onError(e);
            }
        });
    }
    
    /**
     * Sign a message using MPC
     */
    public void signMessage(String message, String address, Callback<String> callback) {
        executor.execute(() -> {
            try {
                JSONObject body = new JSONObject();
                body.put("message", message);
                body.put("address", address);
                
                JSONObject response = makePostRequest("/sign-message", body);
                callback.onSuccess(response.getString("signature"));
            } catch (Exception e) {
                callback.onError(e);
            }
        });
    }
    
    /**
     * Get session key for quick operations
     */
    public void getSessionKey(String address, Callback<SessionKey> callback) {
        executor.execute(() -> {
            try {
                JSONObject response = makeGetRequest("/session/" + address);
                
                SessionKey sessionKey = new SessionKey();
                sessionKey.key = response.getString("key");
                sessionKey.expiresAt = response.getLong("expiresAt");
                
                this.sessionKey = sessionKey.key;
                callback.onSuccess(sessionKey);
            } catch (Exception e) {
                callback.onError(e);
            }
        });
    }
    
    /**
     * Rotate key share for enhanced security
     */
    public void rotateKey(String address, Callback<String> callback) {
        executor.execute(() -> {
            try {
                JSONObject body = new JSONObject();
                body.put("address", address);
                
                JSONObject response = makePostRequest("/rotate", body);
                callback.onSuccess(response.getString("newKeyShare"));
            } catch (Exception e) {
                callback.onError(e);
            }
        });
    }
    
    /**
     * Get wallet information
     */
    public void getWalletInfo(String address, Callback<WalletInfo> callback) {
        executor.execute(() -> {
            try {
                JSONObject response = makeGetRequest("/wallet/" + address);
                
                WalletInfo info = new WalletInfo();
                info.address = response.getString("address");
                info.publicKey = response.getString("publicKey");
                info.createdAt = response.getLong("createdAt");
                info.keyShares = response.getInt("keyShares");
                info.securityLevel = response.getString("securityLevel");
                
                callback.onSuccess(info);
            } catch (Exception e) {
                callback.onError(e);
            }
        });
    }
    
    /**
     * Generate key share locally (for added security)
     */
    public String generateKeyShare() {
        SecureRandom random = new SecureRandom();
        byte[] keyShare = new byte[32];
        random.nextBytes(keyShare);
        return Base64.encodeToString(keyShare, Base64.NO_WRAP);
    }
    
    /**
     * Compute combined public key from shares
     */
    public String computePublicKey(String[] keyShares) throws Exception {
        MessageDigest digest = MessageDigest.getInstance("SHA-256");
        for (String share : keyShares) {
            digest.update(share.getBytes(StandardCharsets.UTF_8));
        }
        byte[] hash = digest.digest();
        return "0x" + Base64.encodeToString(hash, Base64.NO_WRAP);
    }
    
    private JSONObject makePostRequest(String endpoint, JSONObject body) throws Exception {
        URL url = new URL(BASE_URL + endpoint);
        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
        conn.setRequestMethod("POST");
        conn.setRequestProperty("Content-Type", "application/json");
        conn.setDoOutput(true);
        
        conn.getOutputStream().write(body.toString().getBytes(StandardCharsets.UTF_8));
        
        BufferedReader reader = new BufferedReader(
            new InputStreamReader(conn.getInputStream())
        );
        
        StringBuilder response = new StringBuilder();
        String line;
        while ((line = reader.readLine()) != null) {
            response.append(line);
        }
        reader.close();
        
        return new JSONObject(response.toString());
    }
    
    private JSONObject makeGetRequest(String endpoint) throws Exception {
        URL url = new URL(BASE_URL + endpoint);
        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
        conn.setRequestMethod("GET");
        
        BufferedReader reader = new BufferedReader(
            new InputStreamReader(conn.getInputStream())
        );
        
        StringBuilder response = new StringBuilder();
        String line;
        while ((line = reader.readLine()) != null) {
            response.append(line);
        }
        reader.close();
        
        return new JSONObject(response.toString());
    }
    
    // Data classes
    public static class MPCWallet {
        public String address;
        public String publicKey;
        public long createdAt;
        
        public String toJson() {
            try {
                JSONObject obj = new JSONObject();
                obj.put("address", address);
                obj.put("publicKey", publicKey);
                obj.put("createdAt", createdAt);
                return obj.toString();
            } catch (Exception e) {
                return "{}";
            }
        }
    }
    
    public static class SessionKey {
        public String key;
        public long expiresAt;
    }
    
    public static class WalletInfo {
        public String address;
        public String publicKey;
        public long createdAt;
        public int keyShares;
        public String securityLevel;
    }
}

File: browser_extensions/chrome/src/services/mpc-wallet-service.js
/**
 * TigerWallet Browser Extension - MPC Wallet Service
 */

class MPCWalletService {
    constructor() {
        this.apiBase = 'https://api.tigerwallet.com/v1/mpc';
        this.sessionKey = null;
    }

    // Create MPC wallet
    async createWallet(userId) {
        const response = await fetch(`${this.apiBase}/wallet`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ userId })
        });
        if (!response.ok) throw new Error('Failed to create wallet');
        return response.json();
    }

    // Import wallet
    async importWallet(mnemonic, userId) {
        const response = await fetch(`${this.apiBase}/wallet/import`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ mnemonic, userId })
        });
        if (!response.ok) throw new Error('Failed to import wallet');
        return response.json();
    }

    // Get public key
    async getPublicKey(address) {
        const response = await fetch(`${this.apiBase}/publickey/${address}`);
        if (!response.ok) throw new Error('Failed to get public key');
        return response.json();
    }

    // Sign transaction (MPC)
    async signTransaction(txData, address) {
        const response = await fetch(`${this.apiBase}/sign`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ txData, address })
        });
        if (!response.ok) throw new Error('Failed to sign');
        return response.json();
    }

    // Sign message
    async signMessage(message, address) {
        const response = await fetch(`${this.apiBase}/sign-message`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ message, address })
        });
        if (!response.ok) throw new Error('Failed to sign message');
        return response.json();
    }

    // Get session key
    async getSessionKey(address) {
        const response = await fetch(`${this.apiBase}/session/${address}`);
        if (!response.ok) throw new Error('Failed to get session');
        return response.json();
    }

    // Rotate key share
    async rotateKey(address) {
        const response = await fetch(`${this.apiBase}/rotate`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ address })
        });
        if (!response.ok) throw new Error('Failed to rotate key');
        return response.json();
    }

    // Get wallet info
    async getWalletInfo(address) {
        const response = await fetch(`${this.apiBase}/wallet/${address}`);
        if (!response.ok) throw new Error('Failed to get wallet info');
        return response.json();
    }

    // Export key share (requires auth)
    async exportKeyShare(address, authToken) {
        const response = await fetch(`${this.apiBase}/export`, {
            method: 'POST',
            headers: { 
                'Content-Type': 'application/json',
                'Authorization': `Bearer ${authToken}`
            },
            body: JSON.stringify({ address })
        });
        if (!response.ok) throw new Error('Failed to export key');
        return response.json();
    }
}

window.TigerWalletMPCService = new MPCWalletService();
